### PR TITLE
Arnej/add distance threshold

### PR DESCRIPTION
**Pull request by arnej27959**
I confirm that this contribution is made under the terms of the license found in the root directory of this repository's source tree and that I have the authority necessary to make this contribution on behalf of its copyright owner.

@geirst please review

### CODE REVIEW

File: searchlib/src/vespa/searchlib/tensor/distance_functions.h [hunk 1]
```diff
@@ -163,6 +173,11 @@ class GeoDegreesDistance : public DistanceFunction {
         double hav_central_angle = hav_lat + cos(lat_A)*cos(lat_B)*hav_lon;
         return hav_central_angle;
     }
+    double convert_threshold(double threshold) const override {
+        double half_angle = threshold / (2 * 6371.0088);
```
Comment (geirst) [inline searchlib/src/vespa/searchlib/tensor/distance_functions.h:177]: Consider adding a constexpr for the magic constant 6371.0088. Also used in to_rawscore().

File: searchlib/src/tests/tensor/hnsw_index/hnsw_index_test.cpp [hunk 2]
```diff
@@ -144,11 +144,20 @@ class HnswIndexTest : public ::testing::Test {
...
+            auto got_by_docid = index->find_top_k(k, qv, k, 100100.25);
             for (idx = 0; idx < k; ++idx) {
                 EXPECT_EQ(expected_by_docid[idx], got_by_docid[idx].docid);
             }
         }
+        if ((rv.size() > 1) && (rv[0].distance < rv[1].distance)) {
+            double thr = (rv[0].distance + rv[1].distance) * 0.5;
+            auto got_by_docid = index->find_top_k_with_filter(k, qv, *global_filter, k, thr);
+            for (const auto & hit : got_by_docid) {
+                printf("hit docid=%u dist=%g (thr %g)\n", hit.docid, hit.distance, thr);
```
Comment (geirst) [inline searchlib/src/tests/tensor/hnsw_index/hnsw_index_test.cpp:156]: Consider using LOG() instead (that is already setup for this test).

File: searchlib/src/tests/tensor/hnsw_index/hnsw_index_test.cpp [hunk 3]
```diff
@@ -144,11 +144,20 @@ class HnswIndexTest : public ::testing::Test {
         if (exp_hits.size() == k) {
             std::vector<uint32_t> expected_by_docid = exp_hits;
             std::sort(expected_by_docid.begin(), expected_by_docid.end());
-            auto got_by_docid = index->find_top_k(k, qv, k);
+            auto got_by_docid = index->find_top_k(k, qv, k, 100100.25);
             for (idx = 0; idx < k; ++idx) {
                 EXPECT_EQ(expected_by_docid[idx], got_by_docid[idx].docid);
             }
         }
+        if ((rv.size() > 1) && (rv[0].distance < rv[1].distance)) {
```
Comment (geirst) [inline searchlib/src/tests/tensor/hnsw_index/hnsw_index_test.cpp:152]: Currently it is very hard to tell that distance thresholding is actually tested. Consider moving lines 152-160 to a separate function with a descriptive name. And are we sure that all relevant cases are tested when appending this to existing tests? Consider adding an explicit test function to HnswIndexTest if not.

**Review by geirst — APPROVED** [review]
👍

**Comment by arnej27959** [comment]
@geirst PTAL
i think it's enough to just test some simple cases.

**Comment by arnej27959** [comment]
Note: this breaks the query serialization protocol.

**Review by geirst — APPROVED** [review]
👍

**Comment by arnej27959** [comment]
@geirst now added the hack with flag in allow_approximate

**Review by geirst — APPROVED** [review]
👍

**Pull request merged**